evaluator_test.go

```diff
@@ -231,6 +233,81 @@ func TestClusters(t *testing.T) {
 	}))
 }
 
+func TestClustersBoth(t *testing.T) {
+	testEval(t, NewResult("a", "b", "c"), "clusters(one,three)", multiCluster(map[string]Cluster{
+		"a": Cluster{"CLUSTER": []string{"two", "one"}},
+		"b": Cluster{"CLUSTER": []string{"$ALL"}, "ALL": []string{"one"}},
+		"c": Cluster{"CLUSTER": []string{"three"}},
+	}))
+}
+
+func TestClusterQueryCache(t *testing.T) {
+	state := multiCluster(map[string]Cluster{
+		"a": Cluster{"CLUSTER": []string{"two", "one"}},
+		"b": Cluster{"CLUSTER": []string{"$ALL"}, "ALL": []string{"one"}},
+		"c": Cluster{"CLUSTER": []string{"three"}},
+		"d": Cluster{"CLUSTER": []string{"four"}},
+	})
+	state.EnableCachingCQR()
+	state.Query("clusters(one)")
+	expected_length := 1
+	actual_length_of_keys := len(state.cachedCQR.Keys())
+	if actual_length_of_keys != expected_length {
+		t.Errorf("Expected '%d' Key(s) in 'state.cachedCQR', but got %d",
+			expected_length, actual_length_of_keys)
+	}
+}
+
+func TestAddValueToCachedCQR(t *testing.T) {
```

Comment:
Since this isn't a behaviour change, it shouldn't require any extra unit tests. Maybe you want to beef up range-spec because you found some edge cases, but in general a) the existing specs should be good enough to prove correctness, and b) a performance benchmark is proof that you actually made things faster. (Just reading off query times from log would satisfy me - doesn't need to be rigorous. Sounds like you've done that part already.)